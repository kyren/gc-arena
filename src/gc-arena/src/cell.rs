//! GC-aware interior mutability types, enforcing the required write barriers.
//!
//! The core of this module is the [`Mutable<T>`](struct@Mutable) type, which allows
//! distinguishing shared references that allow interior mutability behind [`Gc`](crate::Gc)
//! pointers from those that do not. Combined with custom cell types implementing
//! [`Collect`](crate::Collect) and requiring `&Mutable<Self>` to perform mutation,
//! this provides safe mutation inside GC'd object graphs.
//!
//! The canonical way to safely obtain a `&Mutable<T>` reference is to call
//! [`Gc::mutate`](crate::Gc::mutate) on a [`Gc`](crate::Gc) pointer; this will emit
//! a write barrier, notifying the garbage collected [`Arena`](crate::Arena) that this
//! GC'd object is undergoing mutation and will need to be re-traced.
//!
//! When creating your own types containing GC-aware cells, you'll often want to go from a
//! `&Mutable<Type>` to a `&Mutable<Field>`; the [`#[derive(Mutable)]`](derive@Mutable)
//! macro can then be used to generate safe projection methods.
//!
//! When using third-party types without [`Mutable`](struct@Mutable) projections, the unsafe
//! [`Mutable::assume`] method can be used instead; but be careful to not 'cross' into other
//! [`Gc`](crate::Gc) pointers, as this will violate the invariants of this module.

// Must be declared first to have clean documentation on `Mutable`.
mod mutable;

mod cell;
mod ref_cell;

pub use cell::Cell;
pub use mutable::{DerefNoGc, Mutable};
pub use ref_cell::RefCell;

pub use gc_arena_derive::Mutable;

/// Trait implemented by all GC-aware interior mutability types provided
/// in the [`cell`](self) module.
///
/// Implementing this trait allows for convenient allocation through the
/// [`Gc::allocate_cell`](crate::Gc::allocate_cell) method.
pub trait CellLike {
    /// The type this cell wraps.
    type Target;

    /// Wraps a value inside this cell.
    fn wrap(value: Self::Target) -> Self;
}
