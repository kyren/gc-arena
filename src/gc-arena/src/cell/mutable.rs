use core::ops::Deref;

#[cfg(doc)]
use crate::Gc;

/// A mutable value inside a GC'd object graph.
///
/// This type is only used by reference, and asserts that [`Gc::write_barrier`] was
/// called on the parent [`Gc`] pointer. This allows for further mutation without
/// needing to emit a write barrier again.
#[repr(transparent)]
pub struct Mutable<T: ?Sized>(T);

impl<T: ?Sized> Deref for Mutable<T> {
    type Target = T;

    #[inline(always)]
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: ?Sized> Mutable<T> {
    /// Asserts that the given reference can be safely mutated.
    ///
    /// # Safety
    /// In order to maintain the invariants of the garbage collector, no new [`Gc`] pointers
    /// may be adopted by the referenced value as a result of the interior mutability enabled
    /// by this wrapper, unless [`Gc::write_barrier`] is invoked manually on the parent [`Gc`]
    /// pointer before the next collection is triggered.
    #[inline(always)]
    pub unsafe fn assume(v: &T) -> &Mutable<T> {
        // Safe thanks to #[repr(transparent)]
        core::mem::transmute(v)
    }
}
