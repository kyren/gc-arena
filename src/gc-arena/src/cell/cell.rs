use core::cell::Cell as StdCell;
use core::fmt::{self, Debug};

use crate::{Collect, CollectionContext, Gc, MutationContext};

use super::{CellLike, Mutable};

/// A GC-aware version of `std`'s [`Cell`](StdCell).
pub struct Cell<T: ?Sized>(StdCell<T>);

impl<T: Debug + Copy> Debug for Cell<T> {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        Debug::fmt(&self.0, f)
    }
}

unsafe impl<T: Collect + ?Sized> Collect for Cell<T> {
    fn trace(&self, cc: CollectionContext) {
        // The inner value is never mutated while the GC is running.
        let val = unsafe { &*self.0.as_ptr() };
        val.trace(cc);
    }
}

impl<T> CellLike for Cell<T> {
    type Target = T;

    #[inline]
    fn wrap(value: Self::Target) -> Self {
        Self(StdCell::new(value))
    }
}

impl<T> Cell<T> {
    #[inline(always)]
    pub fn new(value: T) -> Cell<T> {
        Self(StdCell::new(value))
    }

    #[inline(always)]
    pub fn get(&self) -> T
    where
        T: Copy,
    {
        self.0.get()
    }

    #[inline(always)]
    pub fn set(this: &Mutable<Self>, val: T) {
        this.0.set(val)
    }

    #[inline(always)]
    pub fn replace(this: &Mutable<Self>, val: T) -> T {
        this.0.replace(val)
    }

    #[inline(always)]
    pub fn into_inner(self) -> T {
        self.0.into_inner()
    }
}

impl<T: ?Sized> Cell<T> {
    #[inline(always)]
    pub fn get_mut(&mut self) -> &mut T {
        self.0.get_mut()
    }
}

/// Convenience aliases for mutating [`Cell`] methods.
impl<T> Mutable<Cell<T>> {
    #[inline(always)]
    pub fn set(&self, val: T) {
        self.0.set(val)
    }

    #[inline(always)]
    pub fn replace(&self, val: T) -> T {
        self.0.replace(val)
    }
}

/// Convenience aliases for mutating [`Cell`] methods.
impl<'gc, T: 'gc + Collect> Gc<'gc, Cell<T>> {
    #[inline(always)]
    pub fn set(&self, mc: MutationContext<'gc, '_>, val: T) {
        Gc::write_barrier(mc, *self);
        self.0.set(val)
    }

    #[inline(always)]
    pub fn replace(&self, mc: MutationContext<'gc, '_>, val: T) -> T {
        Gc::write_barrier(mc, *self);
        self.0.replace(val)
    }
}
