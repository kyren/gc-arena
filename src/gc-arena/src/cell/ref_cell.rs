use core::cell::{BorrowError, BorrowMutError, Ref, RefCell as StdRefCell, RefMut};
use core::fmt::{self, Debug};

use crate::collect::Collect;
use crate::context::{CollectionContext, MutationContext};
use crate::Gc;

use super::{CellLike, Mutable};

/// A GC-aware [`RefCell`](StdRefCell).
pub struct RefCell<T: ?Sized>(StdRefCell<T>);

impl<T: Debug> Debug for RefCell<T> {
    #[inline(always)]
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        Debug::fmt(&self.0, fmt)
    }
}

unsafe impl<T: Collect + ?Sized> Collect for RefCell<T> {
    #[inline(always)]
    fn trace(&self, cc: CollectionContext) {
        self.0.borrow().trace(cc)
    }
}

impl<T: Collect> CellLike for RefCell<T> {
    type Target = T;

    #[inline(always)]
    fn wrap(value: Self::Target) -> Self {
        Self(StdRefCell::new(value))
    }
}

impl<T> RefCell<T> {
    #[inline(always)]
    pub fn new(value: T) -> Self {
        Self(StdRefCell::new(value))
    }
}

impl<T: ?Sized> RefCell<T> {
    #[inline(always)]
    pub fn as_ptr(&self) -> *mut T {
        self.0.as_ptr()
    }

    #[inline(always)]
    #[track_caller]
    pub fn read(&self) -> Ref<'_, T> {
        self.0.borrow()
    }

    #[inline(always)]
    pub fn try_read(&self) -> Result<Ref<'_, T>, BorrowError> {
        self.0.try_borrow()
    }

    #[inline(always)]
    #[track_caller]
    pub fn write(this: &Mutable<Self>) -> RefMut<'_, T> {
        this.0.borrow_mut()
    }

    #[inline(always)]
    pub fn try_write(this: &Mutable<Self>) -> Result<RefMut<'_, T>, BorrowMutError> {
        this.0.try_borrow_mut()
    }
}

/// Convenience aliases for mutating [`RefCell`] methods.
impl<T: ?Sized> Mutable<RefCell<T>> {
    #[inline(always)]
    #[track_caller]
    pub fn write(&self) -> RefMut<'_, T> {
        self.0.borrow_mut()
    }

    #[inline(always)]
    #[track_caller]
    pub fn try_write(&self) -> Result<RefMut<'_, T>, BorrowMutError> {
        self.0.try_borrow_mut()
    }
}

/// Convenience aliases for mutating [`RefCell`] methods.
impl<'gc, T: 'gc + Collect> Gc<'gc, RefCell<T>> {
    #[inline]
    #[track_caller]
    pub fn write<'a>(&'a self, mc: MutationContext<'gc, '_>) -> RefMut<'a, T> {
        Gc::write_barrier(mc, *self);
        self.0.borrow_mut()
    }

    #[inline]
    #[track_caller]
    pub fn try_write<'a>(
        &'a self,
        mc: MutationContext<'gc, '_>,
    ) -> Result<RefMut<'a, T>, BorrowMutError> {
        Gc::write_barrier(mc, *self);
        self.0.try_borrow_mut()
    }
}
