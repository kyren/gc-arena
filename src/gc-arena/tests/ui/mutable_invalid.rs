#![allow(unused)]

use gc_arena::cell::Mutable;

#[derive(Mutable)]
#[mutable(name = "Dup1", name = "Dup2")]
struct Duplicates {
    #[mutable]
    #[mutable]
    a: i32,
}

#[derive(Mutable)]
#[mutable]
struct InvalidOptions {
    #[mutable(name = "a")]
    a: i32,
    #[mutable = "invalid"]
    b: i32,
}

#[derive(Mutable)]
struct NoMutables(i32);

#[derive(Mutable)]
enum MissingMutName {
    Foo,
    Bar(#[mutable] i32),
}

#[derive(Mutable)]
union Union {
    a: i32,
    #[mutable]
    b: u32,
}

fn main() {}
