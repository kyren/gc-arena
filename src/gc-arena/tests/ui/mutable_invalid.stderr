error: multiple names specified
 --> tests/ui/mutable_invalid.rs:6:26
  |
6 | #[mutable(name = "Dup1", name = "Dup2")]
  |                          ^^^^

error: duplicate `#[mutable(...)]` attribute
 --> tests/ui/mutable_invalid.rs:9:7
  |
9 |     #[mutable]
  |       ^^^^^^^

error: expected non-empty attribute list
  --> tests/ui/mutable_invalid.rs:14:3
   |
14 | #[mutable]
   |   ^^^^^^^

error: unknown option
  --> tests/ui/mutable_invalid.rs:16:15
   |
16 |     #[mutable(name = "a")]
   |               ^^^^

error: expected attribute list
  --> tests/ui/mutable_invalid.rs:18:7
   |
18 |     #[mutable = "invalid"]
   |       ^^^^^^^

error: `#[derive(Mutable)]` requires at least one `#[mutable]` field
  --> tests/ui/mutable_invalid.rs:22:10
   |
22 | #[derive(Mutable)]
   |          ^^^^^^^
   |
   = note: this error originates in the derive macro `Mutable` (in Nightly builds, run with -Z macro-backtrace for more info)

error: `#[derive(Mutable)]` on enums requires `#[mutable(name = "...")]`
  --> tests/ui/mutable_invalid.rs:25:10
   |
25 | #[derive(Mutable)]
   |          ^^^^^^^
   |
   = note: this error originates in the derive macro `Mutable` (in Nightly builds, run with -Z macro-backtrace for more info)

error: `#[derive(Mutable)]` isn't supported on unions
  --> tests/ui/mutable_invalid.rs:31:10
   |
31 | #[derive(Mutable)]
   |          ^^^^^^^
   |
   = note: this error originates in the derive macro `Mutable` (in Nightly builds, run with -Z macro-backtrace for more info)
